vllm/worker/worker.py

```diff
@@ -90,8 +88,9 @@ def profile_num_available_blocks(
         # Calculate the number of blocks that can be allocated with the
         # profiled peak memory.
         torch.cuda.synchronize()
-        peak_memory = torch.cuda.max_memory_allocated()
-        total_gpu_memory = get_gpu_memory()
+        free_gpu_memory, total_gpu_memory = torch.cuda.mem_get_info()
+        peak_memory = total_gpu_memory - free_gpu_memory
```

Comment:
I'm a bit confused. What exactly does the `free_gpu_memory` here measure? Is it the part of memory that pytorch memory allocator does not touch? In other words, is the following diagram correct?

```
|<--------------------------------total GPU memory---------------------------------->|
|<---------------what Pytorch allocator allocates------------->|<--free GPU memory-->|
|<--actually used in execution-->|<--allocator fragmentation-->|<--free GPU memory-->|
```

In this case, `peak_memory` here can greatly over-estimate the memory usage. Is my understanding correct?

Reply:
> In this case, peak_memory here can greatly over-estimate the memory usage. Is my understanding correct?

It depends on the memory fragmentation caused by the allocator. However, I believe the fragmentation should be small, since the most weight/activation tensors are pretty big.

>  is the following diagram correct?

I think one missing part here is the memory that is not allocated/freed using the PyTorch's memory allocator. For example, the memory internally used by cuBLAS or NCCL is not captured by the PyTorch's GPU memory allocator. My hypothesis is the NCCL memory usage is not considered in the current memory profiling which is based on PyTorch.

@zhuohan123

Reply:
Plus, I don't think the PR "over-estimates" the peak memory usage, since I observed that even after this PR, the memory usage goes over the `gpu_memory_utilization`.

> After this PR, the peak memory usage is about 68 GiB for benchmark_throughput.py (with TP=4) and 73 GiB for python benchmarks/benchmark_latency.py --model meta-llama/Llama-2-70b-hf -tp 4 --batch-size 100 --input-len 1024 --output-len 512.

Reply:
I see. It's pretty strange that NCCL uses any extra memory. Since all we use is all_reduce, which should happen in-place and should not use any extra memory.

Reply:
@zhuohan123 

I measure the GPU's peak memory usage (using `torch.cuda.mem_info`), PyTorch reserved memory (`torch.cuda.memory.memory_reserved`) and PyTorch allocated memory (`torch.cuda.memory.memory_allocated`). I measured the memory usage during the initial profiling (i.e., before creating any CUDA graphs).

* llama 7B, TP=1
```
Rank0 peak memory: 13.66, reserved: 13.15, allocated: 12.56
```

1. There's little difference (~500 MiB) between the `peak` and `reserved` memory. I guess this space of memory is spent for some CUDA objects like cuBLAS handles.
2. There's little difference (~600 MiB) between the `reserved` and `allocated` memory. This means the intermediate activation uses ~600 MiB of space.

* llama-70B, TP=8 (before CUDA graph PR)
```
(RayWorkerVllm pid=252359) Rank0 peak memory: 24.57, reserved: 22.63, allocated: 16.07
(RayWorkerVllm pid=252361) Rank2 peak memory: 24.98, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=252360) Rank1 peak memory: 24.98, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=252362) Rank3 peak memory: 24.98, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=252366) Rank7 peak memory: 24.69, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=252364) Rank5 peak memory: 25.10, reserved: 22.88, allocated: 16.07
(RayWorkerVllm pid=252365) Rank6 peak memory: 24.98, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=252363) Rank4 peak memory: 24.98, reserved: 22.75, allocated: 16.07
```

1. There's large difference (~2 GiB) between the `peak` and `reserved` memory. **I guess this is used for cuBLAS handle + NCCL. NCCL might use some GPU memory for the communicator and scratch pad memory. But I'm not sure.**
2. There's large difference (~6.7 GiB) between the `reserved` and `allocated` memory. **I have no idea where this huge difference comes from.**

* llama-70B, TP=8 (after CUDA graph PR)
```
(RayWorkerVllm pid=238494) Rank3 peak memory: 25.66, reserved: 22.63, allocated: 16.07
(RayWorkerVllm pid=238498) Rank7 peak memory: 25.37, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=238493) Rank2 peak memory: 25.79, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=238492) Rank1 peak memory: 25.79, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=238491) Rank0 peak memory: 25.37, reserved: 22.75, allocated: 16.07
(RayWorkerVllm pid=238496) Rank5 peak memory: 25.91, reserved: 22.88, allocated: 16.07
(RayWorkerVllm pid=238497) Rank6 peak memory: 25.91, reserved: 22.88, allocated: 16.07
(RayWorkerVllm pid=238495) Rank4 peak memory: 25.79, reserved: 22.75, allocated: 16.0
```

1. The `peak` memory increases by ~1 GiB. **I guess this is because we use 2 NCCL communicators: one for PyTorch and another for CuPy.**

Reply:
I've visualized the memory usage:

* llama 7B, TP=1
<img width="3346" alt="Screenshot 2023-12-16 at 11 14 03 PM" src="https://github.com/vllm-project/vllm/assets/46394894/e6ed7069-2190-4823-8f25-8e27bd94fe35">

The activation memory is reused after every 

* llama-70B, TP=8
<img width="3247" alt="Screenshot 2023-12-16 at 11 20 10 PM" src="https://github.com/vllm-project/vllm/assets/46394894/b5f492bb-7262-4c06-a040-7796e0f7fc06">

**This shows that the activation memory for all reduce is not reused**